worker: serialize errors if stack getter throws

Trott opened the pull request:
<strike>Remove an apparently unnecessary try/catch from the error
serialization/deserialization internal module used by worker_threads.

(If I'm wrong about this being unnecessary for some reason, I'll switch course and try to add a test and/or an explanatory comment. But it sure seems to effectively be a no-op?)</strike>

Current code that is intended to handle the stack getter throwing is untested. Add a test and adjust code to function as expected.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/20809/

addaleax commented:
The idea behind this line was that `.stack` is implemented as a “hidden” getter on the V8 side, which can call into user code via `Error.prepareStackTrace` (if that is defined), which in turn might throw an error itself. (Probably should have left a comment for that. :))

I don’t think it would be bad to fall back back to the second case (`kSerializedObject`) in such a situation, though – it wouldn’t be very different from a getter on a nested property of the error object throwing an error.

Trott commented:
@addaleax I'm having difficulty finding a way to show this line having an effect. Here's the test I came up with:

```js
'use strict';
const common = require('../common');
const assert = require('assert');
const { Worker } = require('worker_threads');

const w = new Worker(`
  const fn = (err) => {
    if (err.message === 'fhqwhgads')
      throw new Error('come on');
    return 'This is my custom stack trace!';
  };
  Error.prepareStackTrace = fn;
  throw new Error('fhqwhgads');
  `,
  { eval: true }
);
w.on('message', common.mustNotCall());
w.on('error', common.mustCall((err) => {
  console.log(err);
}));
```

But whether I compile with or without the line removed in this PR, the output is the same:

```console
Error [ERR_WORKER_UNSERIALIZABLE_ERROR]: Serializing an uncaught exception failed
    at Worker.[kOnCouldNotSerializeErr] (internal/worker.js:135:24)
    at Worker.[kOnMessage] (internal/worker.js:149:45)
    at MessagePort.Worker.(anonymous function).on (internal/worker.js:84:57)
    at MessagePort.emit (events.js:197:13)
    at MessagePort.onmessage (internal/worker/io.js:62:8)
```

Do I need to revise the test? Or does this show that the line has no effect either way?

addaleax commented:
@Trott Yeah, I get the same results. It looks like V8 tries to call `prepareStackTrace` again if it failed the first time? When adding another try/catch like this, I do get the serialized exception with your test:

```diff
diff --git a/lib/internal/error-serdes.js b/lib/internal/error-serdes.js
index 7b4b416b80c6..bba49c782fff 100644
--- a/lib/internal/error-serdes.js
+++ b/lib/internal/error-serdes.js
@@ -36,7 +36,10 @@ function TryGetAllProperties(object, target = object) {
   Assign(all, TryGetAllProperties(GetPrototypeOf(object), target));
   const keys = GetOwnPropertyNames(object);
   ForEach(keys, (key) => {
-    const descriptor = GetOwnPropertyDescriptor(object, key);
+    let descriptor;
+    try {
+      descriptor = GetOwnPropertyDescriptor(object, key);
+    } catch { return; }
     const getter = descriptor.get;
     if (getter && key !== '__proto__') {
       try {
```

Trott commented:
OK, added the test, and adjusted the code per @addaleax's comment (and added a `Co-Authored-By` for her). PTAL.

CI: https://ci.nodejs.org/job/node-test-pull-request/20899/

addaleax commented:
Thanks!

addaleax commented:
@devsnek You probably want to re-review this :)

BridgeAR commented on lib/internal/error-serdes.js:
```diff
@@ -36,7 +36,10 @@ function TryGetAllProperties(object, target = object) {
+    let descriptor;
+    try {
+      descriptor = GetOwnPropertyDescriptor(object, key);
+    } catch { return; }
```
I am fine with returning undefined but wouldn't it be even better to return the error to the user as well? We could wrap it into something else so it's clear that accessing the property threw (for example by adding a new Node.js error like `ERR_THREW_ON_ACCESS`) which contains the actual error as property.

BridgeAR commented on test/parallel/test-worker-error-stack-getter-throws.js:
```diff
@@ -0,0 +1,22 @@
+  assert.strictEqual(err.stack, undefined);
+  assert.strictEqual(err.message, 'fhqwhgads');
+  assert.strictEqual(err.name, 'Error');
+}));
```
I suggest to change this part to:

```js
w.on('error', common.expectsError({
  stack: undefined,
  message: 'fhqwhgads',
  name: 'Error'
});
```

devsnek commented:
i am confused why we wouldn't just let the error bubble up. the entire state of the resulting object could be *silently* incorrect because of this.

Trott commented:
> i am confused why we wouldn't just let the error bubble up. the entire state of the resulting object could be _silently_ incorrect because of this.

I'm fine with that. How about you, @addaleax?

```console
Error [ERR_WORKER_UNSERIALIZABLE_ERROR]: Serializing an uncaught exception failed
    at Worker.[kOnCouldNotSerializeErr] (internal/worker.js:135:24)
    at Worker.[kOnMessage] (internal/worker.js:149:45)
    at MessagePort.Worker.(anonymous function).on (internal/worker.js:84:57)
    at MessagePort.emit (events.js:197:13)
    at MessagePort.onmessage (internal/worker/io.js:62:8)
```

addaleax commented:
@Trott @devsnek It’s important to me that *some* helpful information about the original error survives, and ideally as much as possible. It’s not clear to me how “Serializing an uncaught exception failed” is better than providing an object without a stack trace (which is, admittedly, already pretty unhelpful).

Trott commented:
> @Trott @devsnek It’s important to me that _some_ helpful information about the original error survives, and ideally as much as possible. It’s not clear to me how “Serializing an uncaught exception failed” is better than providing an object without a stack trace (which is, admittedly, already pretty unhelpful).

So that then brings us to @BridgeAR's suggestion of trying to provide some helpful information when it throws.

Thinking about it more, I would most likely prefer that we emulate as closely as possible what happens outside of workers so that at least people can debug and not be confused that they get one error outside of workers and a different error in a worker. Here's what happens outside of a worker:

```console
/Users/trott/io.js/test/parallel/test-worker-error-stack-getter-throws.js:30
throw new Error('fhqwhgads');
^

Error: fhqwhgads
```

So that would argue for the approach currently here where we return the error with no stack trace at all and no indication of an additional error while creating the stack trace. @devsnek?

devsnek commented:
maybe i'm just missing something obvious here, but why isn't this diff just
```diff
-try { error.stack; } catch {}
```
and nothing else?

Trott commented:
@devsnek That's what it was initially but doing that results in this error:

```console
Error [ERR_WORKER_UNSERIALIZABLE_ERROR]: Serializing an uncaught exception failed
    at Worker.[kOnCouldNotSerializeErr] (internal/worker.js:135:24)
    at Worker.[kOnMessage] (internal/worker.js:149:45)
    at MessagePort.Worker.(anonymous function).on (internal/worker.js:84:57)
    at MessagePort.emit (events.js:197:13)
    at MessagePort.onmessage (internal/worker/io.js:62:8)
```

That's OK, I guess, but it has the following shortcomings:

* The end user has no idea what the original error is.
* The end user has no idea what the additional error thrown by the stack getter was, or even that there was an additional error thrown by the stack getter.
* It is very different from what happens with the same code run outside of a worker.

devsnek commented:
wherever ERR_WORKER_UNSERIALIZABLE_ERROR is thrown we could just re-throw the actual error right?

addaleax commented:
> wherever ERR_WORKER_UNSERIALIZABLE_ERROR is thrown we could just re-throw the actual error right?

@devsnek Can you clarify what you mean by “re-throw”? This is coming from the uncaught exception handler, not from a try/catch.

> to print 'xyz', just like it does in browser.

That’s already happening here as well.

devsnek commented:
ok i see how this code is actually being used now. i would probably just copy what browsers do and ignore the stack property.

https://html.spec.whatwg.org/multipage/workers.html#runtime-script-errors-2

https://repl.it/repls/ForthrightWhoppingParentheses

addaleax commented:
@devsnek I think the stack trace is too valuable to be omitted. I honestly wouldn’t know what to do if I didn’t have it available when debugging the worker tests. :)

Trott commented:
@devsnek I can't tell if your comments are comments/discussion, or if you're -1 on landing this. Can you let me know?

devsnek commented:
@Trott i'm not blocking this

Trott commented:
Landed in 79a3348d142be713c2aa5dbe41524b54cb6bac03
